File: LeetCodeProblemsCMake/Problems/LeetCode3074AppleRedistributionIntoBoxes.cpp
//
// Created by saeed on 12/24/25.
//

#include "LeetCode3074AppleRedistributionIntoBoxes.h"

#include <algorithm>
#include <numeric>
using namespace std;

int LeetCode3074AppleRedistributionIntoBoxes::minimumBoxes(vector<int> &apple, vector<int> &capacity) {
    ranges::sort(capacity, std::greater<>{});
    int sumApples=accumulate(apple.begin(), apple.end(), 0);
    int result{};
    int currentSum{};;
    int length=static_cast<int>(capacity.size());
    for (int i=0; i<length; ++i) {
        currentSum+=capacity[i];
        ++result;
        if (currentSum>=sumApples) {
            return result;
        }

    }
    return result;
}

File: LeetCodeProblemsGo/LeetCode3074AppleRedistributionIntoBoxes/LeetCode3074AppleRedistributionIntoBoxes.go
package LeetCode3074AppleRedistributionIntoBoxes

import (
	"sort"
)

func minimumBoxes(apple []int, capacity []int) int {
	sort.Slice(capacity, func(a, b int) bool {
		return capacity[a] > capacity[b]
	})
	length := len(capacity)
	applesCount := 0
	for _, apple := range apple {
		applesCount += apple
	}
	result := 0
	currentSum := 0
	for i := 0; i < length; i++ {
		currentSum += capacity[i]
		result++
		if currentSum >= applesCount {
			return result
		}
	}
	return result
}
